### PR TITLE
Preserve column metadata during more DataFrame operations

### PR DESCRIPTION
**Pull request by mroeschke**
## Description
Supersedes https://github.com/rapidsai/cudf/pull/15410/, adds a `ColumnAccessor._from_columns_like_self` that will preserve column attributes during DataFrame operations. This can wholly replace `_from_data_like_self`

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/dataframe.py [hunk 1]
```diff
@@ -3091,16 +3094,17 @@ def where(self, cond, other=None, inplace=False):
...
                 out_mask = cudf._lib.null_mask.create_null_mask(
                     len(col),
                     state=cudf._lib.null_mask.MaskState.ALL_NULL,
                 )
-                out[name] = col.set_mask(out_mask)
+                out.append(col.set_mask(out_mask))
 
         return self._mimic_inplace(
-            self._from_data_like_self(out), inplace=inplace
+            self._from_data_like_self(self._data._from_columns_like_self(out)),
```
Comment (vyasr) [inline python/cudf/cudf/core/dataframe.py:3106]: I still see a number of instances of `_from_data_like_self`. Can we use `_from_data` instead consistently now?

Comment (mroeschke) [reply python/cudf/cudf/core/dataframe.py:3106]: > Can we use _from_data instead consistently now?

Getting there but not quite yet due to preserving the `index`. `IndexedFrame._from_data_like_self` is like `_from_data` but preserves the existing index which `IndexedFrame._from_data` doesn't do by default.

File: python/cudf/cudf/core/column_accessor.py [hunk 2]
```diff
@@ -174,6 +175,31 @@ def __repr__(self) -> str:
         )
         return f"{type_info}\n{column_info}"
 
+    def _from_columns_like_self(
+        self, columns: abc.Iterable[ColumnBase], verify: bool = True
+    ):
+        """
+        Return a new ColumnAccessor with columns and the properties of self.
+        """
```
Comment (wence-) [inline python/cudf/cudf/core/column_accessor.py:183]: *nit*: is `verify` documented anywhere?

**Review by wence- — APPROVED** [review]
Minor docstring nit, thanks for the continued refactoring and cleanups!

File: python/cudf/cudf/core/column_accessor.py (hunk 2 above)
Comment (mroeschke) [reply python/cudf/cudf/core/column_accessor.py:183]: This is documented in `ColumnAccessor.__doc__`, but I can also include that in the docstring here

**Pull request merged**